apps/accounts/models/provider_request.py

```diff
@@ -0,0 +1,111 @@
+from django.urls import reverse
+from django_countries.fields import CountryField
+from django.conf import settings
+from django.core.exceptions import ValidationError
+from taggit.managers import TaggableManager
+from apps.greencheck.models import IpAddressField
+from apps.greencheck.validators import validate_ip_range
+from model_utils.models import TimeStampedModel
+from . import Hostingprovider
+
+
+class ProviderRequestStatus(models.TextChoices):
+    PENDING_REVIEW = "Pending review"  # GWF staff needs to verify the request
+    ACCEPTED = "Accepted"  # GWF staff accepted the request
+    REJECTED = "Rejected"  # GWF staff rejected the request (completely)
+    OPEN = "Open"  # GWF staff requested some changes from the provider
+
+
+class ProviderRequest(TimeStampedModel):
+    name = models.CharField(max_length=255)
+    website = models.CharField(max_length=255)
+    description = models.TextField()
+    status = models.CharField(choices=ProviderRequestStatus.choices, max_length=255)
+    created_by = models.ForeignKey(
+        settings.AUTH_USER_MODEL, on_delete=models.CASCADE, null=True
+    )
+
+    def __str__(self):
+        return f"{self.name}"
+
+    def get_absolute_url(self):
+        return reverse("provider_request_detail", args=[str(self.id)])
+
+
+class ProviderRequestSupplier(models.Model):
+    # TODO: add filtering and/or validation to check
+    # that the selected services are offered by the selected supplier.
+
+    supplier = models.ForeignKey(Hostingprovider, on_delete=models.CASCADE)
+    services = TaggableManager(
+        verbose_name="Services used",
+        help_text="Click the services that your organisation uses from the selected supplier. These will be listed in the green web directory.",
+        blank=True,
+    )
+    request = models.ForeignKey(
+        ProviderRequest, on_delete=models.CASCADE, related_name="suppliers"
+    )
+
+    def __str__(self):
+        return f"{self.supplier}"
+
+
+class ProviderRequestLocation(models.Model):
+    city = models.CharField(max_length=255)
+    country = CountryField()
+    services = TaggableManager(
+        verbose_name="Services offered",
+        help_text="Click the services that your organisation offers. These will be listed in the green web directory.",
+        blank=True,
+    )
+    request = models.ForeignKey(ProviderRequest, on_delete=models.CASCADE)
+
+    def __str__(self):
+        return f"{self.request.name} | {self.country}/{self.city}"
+
+
+class ProviderRequestASN(models.Model):
+    asn = models.IntegerField()
+    location = models.ForeignKey(ProviderRequestLocation, on_delete=models.CASCADE)
+
+    def __str__(self):
+        return f"{self.asn}"
+
+
+class ProviderRequestIPRange(models.Model):
+    start = IpAddressField()
+    end = IpAddressField()
+    location = models.ForeignKey(ProviderRequestLocation, on_delete=models.CASCADE)
+
+    def __str__(self):
+        return f"{self.start} - {self.end}"
+
+    def clean(self):
```

Comment:
If we don't already have tests for the IP validation working, it would be nice to see them here. If we do, it's less of a priority.

I think our IPAddressField is a one we built ourselves rather than being from the django framework. If we don't have tests there, then I think it may be worth adding them at that level.

Reply:
There's already a test for the IP validation in another place - the hostingprovider admin (which uses the same code)